docs/libvirt-node-networking.md

```diff
@@ -0,0 +1,298 @@
+cluster, and derive the device from it.
+
+The section on the `Node` from which the IP needs to be taken:
+
+```yaml
+status:
+  addresses:
+  - address: 192.168.121.180
+    type: InternalIP
+```
+
+Based on that IP we can derive the ethernet device which we need. A simple bash
+script like
+```bash
+DEV=$(ip route | grep "src 192.168.121.180" | sed -e "s/.*dev \([^ ]*\).*/\1/g")
+echo $DEV
+eth0
+```
+illustrates that.
+
+Additionally annotations or labels can be considered, to override the detection
+procedure, and directly use a specified device.
+
+## Non-Bridge Scenario (macvlan/macvtap)
+
+```
+  Host               Container                               VM
+| ---------------- | ------------------------------------- | ------------ |
+|           KubeVirt CNI          |                     Libvirt           |
+| ------------------------------- | --------------------------------------| 
+
+  eth0 ----------------macvlan1 -------> macvtap1 -------------> /dev/ethX
+```
+
+virt-handler calls a KubeVirt CNI plugin. The plugin creates a macvlan
+interface, moves it into the container, and attaches the macvlan interface to
+the host interface. The plugin is also responsible for generating/providing MAC
+addresses, if none are provided. The plugin returns the name of the veth
+endpoint in the container, and the MAC address. Libvirt then uses it's `direct`
+attach capability, by creating a macvtap interface, and configures it with the
+MAC address provided.
+
+To check: 
+
+ * Does libvirt need udev in the container or does it the work itself?
+
+Advantage:
+
+ * At the end macvtaps are used, which means that the VM get's the mac address
+   assigned by libvirt
+ * In theory macvlans and macvtaps are faster than bridges.
+ * We can do networking this way, on pretty much every k8s cluster out there,
+   without any specific needs on the host network.
+
+Disadvantage:
+
+ * The VM can't directly communicate with the underlying node, if switches are
+   not configured to mirror traffic. However, a non intrusive work-around will
+   be scetched below.
+
+### How can the node and the nested node communicate with each other?
+
+In case of `macvlan` and `macvtap`, normally hosts and guests/containers can't
+communicate with each other. However, we can add an additional `macvlan`
+interface, bound to the same host interfaces like all other `macvlan` devices.
+We can then route all guest traffic which needs to reach thoe node via it's
+node IP through that `macvlan` device, to allow host to guest and guest to host
+communication ([1], [2]).
+
+To illustrate that, let's suppose that we have the follwoing scenario:
+
+```
+| host                     | guest                    |
+|--------------------------|--------------------------|
+| eth0 (192.168.200.1)     | macvtap1 (192.168.200.3) |
+| macvlan0 (192.168.200.2) |                          |
+```
+
+`macvlan0` and `macvtap1` are both attache to `eth0`. Therefore, `macvlan0` and
```

Comment:
The other diagram further above says that the chain is eth0, macvlan, macvtap ?
Here you say it's eth, macvlan and eth0, macvtap.